### ISSUE TITLE
Live Data not shown properly HEATIROOMCONTROLLER2

### ISSUE THREAD

**Issue by matlab22** [mschlenstedt/LoxBerry-Plugin-Stats4Lox-NG#145]
The "Intelligente Raumregelung (Irc2)" is not properly read. H is not showing the correct value (10 instead of 43.7), teta_c is missing and a lot of entries are not interpreted at all.

<img width="1182" height="194" alt="Image" src="https://github.com/user-attachments/assets/31ee1bed-f7f8-41c1-8503-2f17f61b4b99" />

<img width="905" height="1675" alt="Image" src="https://github.com/user-attachments/assets/146e78e9-5748-4a38-992a-943f5f482345" />

**Comment by mschlenstedt** [comment]
Partially addressed in [`3466b7e`](https://github.com/mschlenstedt/LoxBerry-Plugin-Stats4Lox-NG/commit/3466b7e) - **leaving this open**, because the second half of your report is not fixed yet.

**What is fixed:** `HEATIROOMCONTROLLER2` was missing from the element definitions entirely, which is why no output was described properly. The definitions are regenerated from Loxone Config 17 and now contain the block as "Intelligente Raumregelung", ControlType 469, with all 18 outputs.

That also explains your missing `teta_c`: Loxone names that output `ϑt` with a Greek theta (old label `AQt`, description "Zieltemperatur") - a literal lookup never found it. Related to that, 97 of the description keys turned out to be printf patterns (`H%d`, `Q%d`); the lookup now matches those as well, with the exact match keeping priority.

**What is not fixed:** the wrong value you report - `H` showing 10 instead of 43.7. That is not a naming problem. `Stats4Lox.pm::msget_value` "repairs" Loxone's JSON with regular expressions before decoding it:

```perl
$respvalue_filtered =~ s\.*\"value\": \"([-\d\.]+).*/$1/;
$rawdata =~ s\"value\": ([^\"]+[a-zA-Z]+[^\"]+)\}/\"value\": \"$1\"\}/g;
```

On a block with many outputs those patterns can match the wrong occurrence, which fits your symptom exactly. The fix is to parse the response as real JSON instead, and that is still to be done.

If you can attach the raw response of `/jdev/sps/io/<uuid>/all` for your Irc2 block, that would pin it down precisely.

**Comment by mschlenstedt** [comment]
Please try upcoming Version 1.0 of the plugin. The bug should be fixed.

**Comment by mschlenstedt** [comment]
The remaining half is fixed as well, in [`4470c4f`](https://github.com/mschlenstedt/LoxBerry-Plugin-Stats4Lox-NG/commit/4470c4f) - so this should be complete now.

`msget_value` no longer repairs Loxone's JSON with regular expressions. Verified against a live Miniserver with Loxone Config 17: the response is valid JSON, so the "repair" could only ever corrupt it.

Three defects were in there, all silent:

**1. A failed match reused the previous block's number.**

```perl
my $value = $respjson->{LL}->{value};
$value =~ m/([-+]?[0-9]*\.?[0-9]+)/;   # no check whether this matched
$value = $1+0;
```

`LL.value` may hold no number at all - text blocks answer with `""` or with `"TextValue, not Implemented"`. On a failed match `$1` still held the capture of the **previous** block. Demonstrated: after a block reporting `43.7 °C`, the next two blocks with `""` and `"TextValue, not Implemented"` both reported 43.7. That is very likely what you saw as a wrong value.

**2. The unit was never populated** - it was taken from `$2`, but the pattern has a single group.

**3. The retry without `/all`** fired far more often than intended, because the trigger numified the whole JSON document, which yields 0 for any response. That retry writes on some block types - see #143.

Both are handled by a new `parse_loxone_value()`, which returns number and unit and passes a value through as text when it holds no number instead of inventing one.

Checked against your block type on a live Miniserver: all nine `HeatIRoomController2` blocks now deliver their target temperature output with plausible per-room values (24 °C, 21 °C, 18.5 °C), 41 entries per block, none without a value.

On `teta_c`: the output is named `ϑt` by Loxone, with a Greek theta (old label `AQt`, "Zieltemperatur"). It was never missing from the response - the element definitions simply had no entry for `HEATIROOMCONTROLLER2` at all, which is fixed in [`3466b7e`](https://github.com/mschlenstedt/LoxBerry-Plugin-Stats4Lox-NG/commit/3466b7e).

Closing this. If `H` still shows a wrong value on your system after the update, please reopen with the raw response of `/jdev/sps/io/<uuid>/all` - that would then be a different cause.

**Comment by matlab22** [comment]
It looks good, but I might created / have another issue. There were older controllers with leftovers. So I deleted the old one and kept the date and linked the new to the old date. Everything looks fine, but I can not alter stats.json anymore.Whatever I modify in this mask

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/a502e27f-c2dc-4f16-9956-887fea0eb82d" />

is lost after closing the window. It looks like everything is affected. Is this still working on your site?

**Comment by mschlenstedt** [comment]
That looks like the bug fixed in 1.0.1, but it cannot be the same one - and I need one piece of information from you to say which it is.

**What 1.0.1 fixed** (`dc38090`): on a **fresh** installation `stats.json` was shipped as `{}`. `updatestat` requires a `loxone` array in it and no code path ever created one, so it refused *every* save with HTTP 500. Your installation has entries, so that path is already past - unless you are still on 1.0.0. Worth ruling out first: which version does the plugin show?

**What 1.0.1 did not fix** is why this is invisible. `js/loxone.js` posts the change and had no `fail` handler, so a refused save looked exactly like an accepted one: the dialog keeps the values on screen, and only the next reload shows that nothing was written. `updatestat` refuses for six different reasons - `stats.json` unreadable or unlockable, outputs without matching keys and labels, a missing name/uuid/msno/active - and every one of them ended in the same silence. That is what made your report look like the old bug.

**The reason is in the response.** Two ways to read it, either is enough:

- **Log manager**, the plugin's `ajax` log. A refused save is written there as `<reason> - responding with HTTP 500`.
- **Browser**, F12 -> Network tab, then change something in the dialog. The request to `ajax.cgi` answers `500` with `{"error": "..."}` in the body.

Since you say everything is affected and not just the block you re-linked, my first guess is `stats.json could not be opened or locked` - that would fit a `stats.json` that is no longer valid JSON, which is easy to end up with when entries are re-linked by hand. But the message will name it exactly, so please post it.

The silence itself is fixed for the next version: instead of pretending the value was stored, the dialog now shows the reason the backend gave, in red, right in the settings mask.

**Comment by matlab22** [comment]
I reduced the stats.json

```
{
   "loxone" : [
   ],
   "mqtt" : {
      "subscriptions" : [
         {
            "extractNumbers" : false,
            "id" : "ems-esp/boiler_data"
         },
         {
            "id" : "ems-esp/mixer_data"
         },
         {
            "id" : "ems-esp/thermostat_data"
         },
         {
            "id" : "ems-esp/custom_data"
         }
      ]
   }
}
```

and still could not save anything. I now patched it to have the latest error message and got the 500 error
<img width="300" alt="Image" src="https://github.com/user-attachments/assets/c5dfaa00-830a-40c8-95c1-60e4ce2bd5e1" />

**Comment by matlab22** [comment]
I solved the issue.
defaultDashboard.json was defect. It ended with a lot of NUL:

<img width="100" alt="Image" src="https://github.com/user-attachments/assets/eec4bf8d-e664-44d6-90dc-7883a6dea36f" />

I restored a backup and it is now working.

```
         }
      }
   ],
   "refresh" : "5s",
   "schemaVersion" : 27,
   "style" : "dark",
   "tags" : [],
   "templating" : {
      "list" : []
   },
   "time" : {
      "from" : "now-6h",
      "to" : "now"
   },
   "timepicker" : {},
   "timezone" : "",
   "title" : "LoxBerry Stats4Lox",
   "uid" : "b59c4ca6-895c-4cb4-9822-af0edd264ba8",
   "version" : 0
}
```

I would not bother too much about this. Could be my fault I was editing the stats.json while the plugin still was running.
